disnake/enums.py

```diff
@@ -72,324 +85,1357 @@
 )
 
 
-class _EnumValueBase(NamedTuple):
-    if TYPE_CHECKING:
-        _cls_name: ClassVar[str]
+_T = TypeVar("_T", bound="EnumMeta")
 
-    name: str
-    value: Any
 
-    def __repr__(self) -> str:
-        return f"<{self._cls_name}.{self.name}: {self.value!r}>"
+def _is_descriptor(obj):
+    return hasattr(obj, "__get__") or hasattr(obj, "__set__") or hasattr(obj, "__delete__")
 
-    def __str__(self) -> str:
-        return f"{self._cls_name}.{self.name}"
 
+class _EnumDict(Dict[str, Any]):
+    def __init__(self, base: Type[Any]):
+        # We explicitly take a base in __init__, unlike default Enums, to more easily
+        # enforce proper value typing. This should ensure better performance, as e.g.
+        # `Enum[int].one + 3 == 4` is more performant than `Enum[Any].one.value + 3 == 4`,
+        # as this saves a lookup. We do lose some functionality here, but it's probably
+        # worth the cost; especially considering all current enums have monotyped values.
+        super().__init__()
+        self.base = base
+        self.member_map: Dict[str, Any] = {}
+        self.value_map: Dict[Any, str] = {}
 
-@total_ordering
-class _EnumValueComparable(_EnumValueBase):
-    def __eq__(self, other: object) -> bool:
-        return isinstance(other, self.__class__) and self.value == other.value
+    def __getitem__(self, name: str) -> Any:
+        try:
+            return super().__getitem__(name)
+        except KeyError:
+            return self.value_map[name]
 
-    def __lt__(self, other: object) -> bool:
-        return isinstance(other, self.__class__) and self.value < other.value
+    def __setitem__(self, name: str, value: Any) -> None:
+        if name in {"mro", ""}:  # illegal names defined in original python enums
+            raise ValueError(f"Invalid Enum member name: {name}")
 
+        if name.startswith("_") or _is_descriptor(value):
+            super().__setitem__(name, value)
+            return
 
-def _create_value_cls(name: str, comparable: bool) -> Type[_EnumValueBase]:
-    parent = _EnumValueComparable if comparable else _EnumValueBase
-    return type(f"{parent.__name__}_{name}", (parent,), {"_cls_name": name})  # type: ignore
+        if not isinstance(value, self.base):
+            raise TypeError(
+                f"Member {name} must be of type {self.base.__name__}, got {type(value).__name__}"
+            )
 
+        if name in self.member_map:
+            raise TypeError(f"Attempted to reuse key: '{name!r}'")
```

Comment:
A more descriptive error message would be nice

Reply:
This error message is the same as that of the vanilla python enums. But yeah, it could definitely be a little clearer. I have changed it to `f"Cannot create multiple members with the same name: {name!r}"` as part of fce05efc82105113ab5bb08463c74594b5cdf3d2.